server/src/server.rs

```diff
@@ -258,15 +258,32 @@ impl ServerStreamCallbacks {
             output.len()
         );
 
-        unsafe {
-            if self.input_frame_size != 0 {
+        if self.input_frame_size != 0 {
+            if input.len() > self.shm.get_size() {
+                debug!(
+                    "bad input size: input={} shm={}",
+                    input.len(),
+                    self.shm.get_size()
+                );
+                return -1;
```

Comment:
Need to use `CUBEB_ERROR` explicitly and fire an error callback here?

It seems the stream will start draining, and then in the next callback, it will notify users the stream is drained via the state callback, but users don't know what happens.

Reply:
I'll switch to `CUBEB_ERROR`, that is clearer.

The state callback with `CUBEB_STATE_ERROR` is expected to be handled by the cubeb backend after `data_callback` returns an error.  The CoreAudio backend handles this correctly already, PulseAudio handles the error but doesn't trigger the state callback (I think, need to verify), Windows doesn't handle failed data_callbacks yet: https://github.com/mozilla/cubeb/issues/553 - I'll follow up the remaining cubeb issues on that side.